import asyncio
import logging
import numpy as np
import grpc
import sys
import os

# –î–æ–±–∞–≤–ª—è–µ–º –∫–æ—Ä–Ω–µ–≤—É—é –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é –≤ –ø—É—Ç—å –¥–ª—è –∏–º–ø–æ—Ä—Ç–∞
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import streaming_pb2
import streaming_pb2_grpc
from audio_player import AudioPlayer
from rich.console import Console

# –ù–∞—Å—Ç—Ä–æ–π–∫–∞ –ª–æ–≥–∏—Ä–æ–≤–∞–Ω–∏—è
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

console = Console()

class GrpcClient:
    """gRPC –∫–ª–∏–µ–Ω—Ç –¥–ª—è —Å—Ç—Ä–∏–º–∏–Ω–≥–∞ –∞—É–¥–∏–æ –∏ —Ç–µ–∫—Å—Ç–∞"""
    
    def __init__(self, server_address: str = "localhost:50051"):
        self.server_address = server_address
        self.audio_player = AudioPlayer(sample_rate=48000)
        self.channel = None
        self.stub = None
    
    async def connect(self):
        """–ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ gRPC —Å–µ—Ä–≤–µ—Ä—É"""
        try:
            # –°–æ–∑–¥–∞–µ–º –Ω–µ–±–ª–æ–∫–∏—Ä—É—é—â–∏–π –∫–∞–Ω–∞–ª
            self.channel = grpc.aio.insecure_channel(self.server_address)
            self.stub = streaming_pb2_grpc.StreamingServiceStub(self.channel)
            
            console.print(f"[bold green]‚úÖ –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ gRPC —Å–µ—Ä–≤–µ—Ä—É {self.server_address} —É—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–æ[/bold green]")
            return True
            
        except Exception as e:
            console.print(f"[bold red]‚ùå –û—à–∏–±–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ —Å–µ—Ä–≤–µ—Ä—É: {e}[/bold red]")
            return False
    
    async def disconnect(self):
        """–û—Ç–∫–ª—é—á–µ–Ω–∏–µ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞"""
        if self.channel:
            await self.channel.close()
            console.print("[bold yellow]üîå –û—Ç–∫–ª—é—á–µ–Ω–æ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞[/bold yellow]")
    
    async def stream_audio(self, prompt: str, screenshot_base64: str = None, screen_info: dict = None, hardware_id: str = None):
        """
        –ó–∞–ø—É—Å–∫–∞–µ—Ç —Å—Ç—Ä–∏–º–∏–Ω–≥ –∞—É–¥–∏–æ –∏ —Ç–µ–∫—Å—Ç–∞.
        –≠—Ç–∞ —Ñ—É–Ω–∫—Ü–∏—è —è–≤–ª—è–µ—Ç—Å—è –∞—Å–∏–Ω—Ö—Ä–æ–Ω–Ω—ã–º –≥–µ–Ω–µ—Ä–∞—Ç–æ—Ä–æ–º, –∫–æ—Ç–æ—Ä—ã–π —Å–Ω–∞—á–∞–ª–∞ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç 
        –æ–±—ä–µ–∫—Ç –≤—ã–∑–æ–≤–∞ (–¥–ª—è –≤–æ–∑–º–æ–∂–Ω–æ—Å—Ç–∏ –æ—Ç–º–µ–Ω—ã), –∞ –∑–∞—Ç–µ–º –Ω–∏—á–µ–≥–æ –Ω–µ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç, 
        –ø–æ—Å–∫–æ–ª—å–∫—É –æ–±—Ä–∞–±–æ—Ç–∫–∞ –ø—Ä–æ–∏—Å—Ö–æ–¥–∏—Ç –≤–Ω—É—Ç—Ä–∏.
        """
        if not self.stub:
            console.print("[bold red]‚ùå –ù–µ –ø–æ–¥–∫–ª—é—á–µ–Ω –∫ —Å–µ—Ä–≤–µ—Ä—É[/bold red]")
            return
        
        call = None
        try:
            console.print(f"[bold yellow]üöÄ –ó–∞–ø—É—Å–∫ gRPC —Å—Ç—Ä–∏–º–∏–Ω–≥–∞ –¥–ª—è: {prompt}[/bold yellow]")
            
            if screenshot_base64:
                console.print(f"[bold blue]üì∏ –û—Ç–ø—Ä–∞–≤–ª—è—é —Å–∫—Ä–∏–Ω—à–æ—Ç: {screen_info.get('width', 0)}x{screen_info.get('height', 0)} –ø–∏–∫—Å–µ–ª–µ–π[/bold blue]")
            
            if hardware_id:
                console.print(f"[bold blue]üÜî –û—Ç–ø—Ä–∞–≤–ª—è—é Hardware ID: {hardware_id[:16]}...[/bold blue]")
            
            self.audio_player.start_playback()
            
            request = streaming_pb2.StreamRequest(
                prompt=prompt,
                screenshot=screenshot_base64 if screenshot_base64 else "",
                screen_width=screen_info.get('width', 0) if screen_info else 0,
                screen_height=screen_info.get('height', 0) if screen_info else 0,
                hardware_id=hardware_id if hardware_id else ""
            )
            
            call = self.stub.StreamAudio(request)
            
            # –°—Ä–∞–∑—É –≤–æ–∑–≤—Ä–∞—â–∞–µ–º –æ–±—ä–µ–∫—Ç –≤—ã–∑–æ–≤–∞, —á—Ç–æ–±—ã main –º–æ–≥ –µ–≥–æ –æ—Ç–º–µ–Ω–∏—Ç—å
            yield call
            
            async for response in call:
                if response.HasField('text_chunk'):
                    console.print(f"[green]üìÑ –¢–µ–∫—Å—Ç: {response.text_chunk}[/green]")
                
                elif response.HasField('audio_chunk'):
                    audio_chunk = np.frombuffer(
                        response.audio_chunk.audio_data, 
                        dtype=response.audio_chunk.dtype
                    ).reshape(response.audio_chunk.shape)
                    self.audio_player.add_chunk(audio_chunk)
                
                elif response.HasField('end_message'):
                    console.print(f"[bold green]‚úÖ {response.end_message}[/bold green]")
                    break
                
                elif response.HasField('error_message'):
                    console.print(f"[bold red]‚ùå –û—à–∏–±–∫–∞ –æ—Ç —Å–µ—Ä–≤–µ—Ä–∞: {response.error_message}[/bold red]")
                    break
            
            self.audio_player.wait_for_queue_empty()
            
        except grpc.aio.AioRpcError as e:
            if e.code() == grpc.StatusCode.CANCELLED:
                console.print("[bold yellow]‚ö†Ô∏è –°—Ç—Ä–∏–º–∏–Ω–≥ –æ—Ç–º–µ–Ω–µ–Ω –∫–ª–∏–µ–Ω—Ç–æ–º[/bold yellow]")
            else:
                console.print(f"[bold red]‚ùå gRPC –æ—à–∏–±–∫–∞: {e.details()}[/bold red]")
        except Exception as e:
            console.print(f"[bold red]‚ùå –ü—Ä–æ–∏–∑–æ—à–ª–∞ –Ω–µ–ø—Ä–µ–¥–≤–∏–¥–µ–Ω–Ω–∞—è –æ—à–∏–±–∫–∞ –≤ —Å—Ç—Ä–∏–º–∏–Ω–≥–µ: {e}[/bold red]")
        finally:
            if self.audio_player.is_playing:
                self.audio_player.stop_playback()
            # –£–±–µ–∂–¥–∞–µ–º—Å—è, —á—Ç–æ call –∑–∞–≤–µ—Ä—à–µ–Ω, –µ—Å–ª–∏ –æ–Ω –±—ã–ª —Å–æ–∑–¥–∞–Ω
            if call and not call.done():
                call.cancel()

async def main():
    """–û—Å–Ω–æ–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –∫–ª–∏–µ–Ω—Ç–∞"""
    client = GrpcClient()
    
    try:
        # –ü–æ–¥–∫–ª—é—á–∞–µ–º—Å—è –∫ —Å–µ—Ä–≤–µ—Ä—É
        if not await client.connect():
            return
        
        # –û—Å–Ω–æ–≤–Ω–æ–π —Ü–∏–∫–ª
        while True:
            prompt = console.input("[bold cyan]üé§ –í–≤–µ–¥–∏—Ç–µ –ø—Ä–æ–º–ø—Ç (–∏–ª–∏ 'quit'): [/bold cyan]")
            if prompt.lower() == 'quit':
                break
            
            # –ó–∞–ø—É—Å–∫–∞–µ–º —Å—Ç—Ä–∏–º–∏–Ω–≥
            await client.stream_audio(prompt)
    
    except KeyboardInterrupt:
        console.print("\n[bold yellow]üëã –í—ã—Ö–æ–¥...[/bold yellow]")
    except Exception as e:
        console.print(f"[bold red]‚ùå –ü—Ä–æ–∏–∑–æ—à–ª–∞ –Ω–µ–ø—Ä–µ–¥–≤–∏–¥–µ–Ω–Ω–∞—è –æ—à–∏–±–∫–∞: {e}[/bold red]")
    finally:
        await client.disconnect()
        logger.info("gRPC –∫–ª–∏–µ–Ω—Ç –∑–∞–≤–µ—Ä—à–∏–ª —Ä–∞–±–æ—Ç—É.")

if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        console.print("\n[bold yellow]üëã –í—ã—Ö–æ–¥...[/bold yellow]")
